source/Plugins.cpp

```diff
@@ -38,14 +64,135 @@ namespace {
 				continue;
 
 			for(const DataNode &child : node)
-				if(child.Size() == 2)
+				if(child.Size() == 3)
 				{
+					lock_guard<mutex> guard(pluginsMutex);
 					auto *plugin = plugins.Get(child.Token(0));
 					plugin->enabled = child.Value(1);
 					plugin->currentState = child.Value(1);
+					plugin->version = child.Token(2);
 				}
 		}
 	}
+
+	bool oldNetworkActivity = false;
+
+	mutex activePluginsMutex;
+	set<string> activePlugins;
+
+	// The maximum size of a plugin in bytes, this will be 1 GB.
+	const int MAX_DOWNLOAD_SIZE = 1000000000;
+
+	// Copy an entry from one archive to the other
+	void CopyData(struct archive *ar, struct archive *aw)
+	{
+		int retVal;
+		const void *buff;
+		size_t size;
+		int64_t offset;
+
+		while(true)
+		{
+			retVal = archive_read_data_block(ar, &buff, &size, &offset);
+			if(retVal == ARCHIVE_EOF)
+				return;
+			if(retVal != ARCHIVE_OK)
+				return;
+			if(archive_write_data_block(aw, buff, size, offset) != ARCHIVE_OK)
+				return;
+		}
+	}
+
+	bool ExtractZIP(std::string &filename, const std::string &destination, const std::string &expectedName)
+	{
+		int flags = ARCHIVE_EXTRACT_TIME;
+		flags |= ARCHIVE_EXTRACT_PERM;
+		flags |= ARCHIVE_EXTRACT_ACL;
+		flags |= ARCHIVE_EXTRACT_FFLAGS;
+
+		// Create the handles for reading/writing
+		archive *read = archive_read_new();
+		archive *ext = archive_write_disk_new();
+		archive_write_disk_set_options(ext, flags);
+		archive_read_support_format_all(read);
+		if(!filename.empty() && filename.front() == '-')
+			filename.clear();
+		if(archive_read_open_filename(read, filename.c_str(), 10240))
+			return false;
+
+		// Check if this plugin has the right head folder name
+		archive_entry *entry;
+		archive_read_next_header(read, &entry);
+		string firstEntry = archive_entry_pathname(entry);
+		firstEntry = firstEntry.substr(0, firstEntry.find("/")) + "/";
+		bool fitsExpected = firstEntry == (expectedName);
+		archive_read_data_skip(read);
+
+		// Check if this plugin has a head folder, if not create one in the destination
+		archive_read_next_header(read, &entry);
+		string secondEntry = archive_entry_pathname(entry);
+		bool hasHeadFolder = secondEntry.find(firstEntry) != std::string::npos;
+		if(!hasHeadFolder)
+#if defined(_WIN32)
+			_wmkdir(Utf8::ToUTF16(destination + expectedName).c_str());
+#else
+			mkdir((destination + expectedName).c_str(), 0777);
+#endif
```

Comment:
Don't use unsanitized names as file paths. Ever.

There is nothing stopping someone from putting some path traversal characters into their plugin name. If I call my plugin `../../../../ProgramData/Something/` it's  going to download the plugin into some place it really shouldn't.

Reply:
I would expect the official ES plugin collection to be guarded against that but I see your problem, so any idea how  would check that?

Reply:
I could just check if the name contains ".." right?

Reply:
I added a check not in this function but in the place that calls it, as this function may be called with a ".."